### PR TITLE
core: let OnCalendar= timer units expire during suspend (#8231)

### PR DESCRIPTION
**Pull request by sourcejedi**
If clock jumps forward, some timer units may have expired.  Let them
trigger before we recalculate their timers, because recalculating the timers
will clear the pending timer event.  (If you wonder why we recalculate the
timers, it's to handle *backwards* timejumps).

Implement this by lowering the priority of m->time_change_event_source.

Previously, these two things were both set at default priority so their
ordering, and hence our behaviour, was formally undefined.  I suspect the
sd-event implementation favoured the event source that was created first.

I'd be happier with this if I could avoid numeric priorities. But I didn't
find another way to do this, short of pushing OnCalendar= type events down
in to sd-event as a second type of timer event.

This fixes the reproducer in #8231, such that `atop-rotate.service` can be seen
being triggered by `atop-rotate.timer` as soon as the system wakes from suspend.

Motivations:

* Rotate monitoring data from Atop into files which are named per-day.
  Fedora currently implements this with a cron job that runs at midnight,
  but that didn't handle suspend correctly either.

* unbound-anchor.timer on Fedora, is used to update DNSSEC "root trust
  anchor" daily, before the TTL expires.  It uses OnCalendar=daily
  AccuracySec=24h.  Which is a bit suspect because the TTL is 2 days, but I
  think it has the right general idea.

* In general, we can expect OnCalendar= to be used for the same cases as
  cron.daily and cron.monthly.  Which use anacron to keep track of jobs
  which should be run even if the system was down at the time.

  Timers which are configured to run more frequently than that, are
  unlikely to mind if they get run slightly more often that the writer
  realized, relative to the amount of time the system was really running.

* User report
  https://unix.stackexchange.com/questions/351829/systemd-timer-that-expired-while-suspended

### CODE REVIEW

**Comment by sourcejedi** [comment]
I'd be happier with this if I could avoid numeric priorities.  But I didn't find another way to do this, short of pushing `OnCalendar=` type events down in to sd-event as a second type of timer event.

**Comment by keszybz** [comment]
A bit of a hack, but looks OK. @poettering ?

**Comment by sourcejedi** [comment]
Ug, sorry.  I was getting tired, this is a pain to test, and I really wanted `atop` working properly to get memory usage history the next time my laptop [thrashes](https://en.wikipedia.org/wiki/Thrashing_(computer_science)) itself into non-responsiveness.

I think the timer unit code still looks very suspicious for timejumps which go backwards.

So maybe the correct solution is to fiddle the `base` time used in timer_enter_waiting().  And when time jumps backwards, if that puts the last trigger time in the future, then ignore the last trigger time like we do for persistent files.

**Comment by sourcejedi** [comment]
Another argument for this is the user report, which suggested that systemd behaves correctly if the timer unit had already triggered once.

**Comment by poettering** [comment]
> Previously, these two things were both set at default priority so their
ordering, and hence our behaviour, was formally undefined.  I suspect the
sd-event implementation favoured the event source that was created first.

sd-event favours the event that the kernel told us about first. But given that epoll itself doesn't do any ordering between events this doesn't actually give us too much: if multiple events are triggered in one epoll_wait() cycle then the order is really undefined...

**Comment by poettering** [comment]
I think this change makes a lot of sense, but could we add some more commenting about this? Specifically, I'd like to see an explicit comment here that says something like "Do note that timer event sources of .timer units are scheduled at SD_EVENT_PRIORITY_NORMAL, and hence time units that already triggered will be processed before we process the time change. This is particularly relevant during system suspend cycles, which will generally result in a time change event, but possibly also in triggered time units. There, we want to process timers triggered during suspend first, before we reschedule them for the new time." or so. And then, a similar (though maybe briefer) note should be added to the code in timer units that allocates the timer event sources, and points to the fact timer units have a lower priority than time change events.

Managing ordering through numeric priorities is brittle, hence we really should make sure to document the relationships at the right places. I tried to do that at some places already, and we haven't been very good at this originally, but let's please be careful about this for the future, to make this more robust for future changes.

**Comment by keszybz** [comment]
@sourcejedi can you update this?

**Comment by sourcejedi** [comment]
I can push it for review if you like.  The effect of the last commit is not tested.  I'm travelling tomorrow, hoping my train isn't cancelled because of UK snowfall :), don't have much time at the moment.

**Comment by sourcejedi** [comment]
Pushed.  Correction, the untested effect is "core: don't freeze OnCalendar= timer units when the clock goes back a lot".

You can ignore the fighty footnote in the last commit, I already filed [downstream](https://bugzilla.redhat.com/show_bug.cgi?id=1550259).  So probably just a question of whether we want to fix the message to say systemd-timesynd.service is "enabled", or whether we change the behaviour somehow so it's not wildly inconsistent with the word "active".

File: src/timedate/timedatectl.c [hunk 1]
```diff
@@ -124,10 +124,10 @@ static void print_status_info(const StatusInfo *i) {
         else
                 tzset();
 
-        printf("                       Time zone: %s (%s)\n"
-               "       System clock synchronized: %s\n"
-               "systemd-timesyncd.service active: %s\n"
-               "                 RTC in local TZ: %s\n",
+        printf("                        Time zone: %s (%s)\n"
+               "        System clock synchronized: %s\n"
+               "systemd-timesyncd.service enabled: %s\n"
```
Comment (keszybz) [inline src/timedate/timedatectl.c]: This change looks correct, even if doesn't fix everything. This value comes from https://github.com/sourcejedi/systemd/blob/828553eb0d2d4e05eb074ba9cb9aba2ac6179e05/src/timedate/timedated.c#L179 `use_ntp = STR_IN_SET(s, "enabled", "enabled-runtime")`, so "enabled" is certainly more appropriate than "active".

**Comment by keszybz** [comment]
This looks all good, let's just wait to see if CI passes. If you care, you can update the message on the last commit.

**Comment by sourcejedi** [comment]
Dropped the whole WIP commit.  I aim to test the behaviour of commit #3 at some later date.

Sorry - at the time I pushed the branch, I'd forgotten the WIP commit was there.  It was at least missing the equivalent fix in the manpage.  The patch is still available [here](https://github.com/sourcejedi/systemd/commit/64cab1d6b9ec465449c75839bfb66049df664a7b) just in case someone wanted to look at it.

**Pull request merged**